### DISCUSSION TITLE
Using Async generators in the startup event

### DISCUSSION THREAD

**Discussion by m4nuC** [fastapi/fastapi#6989] [Questions]
Hi,

I am running into a strange issue. I am using Aikafka to run a consumer which I start in the startup event. It looks like this:
```
async def consume():
    consumer = AIOKafkaConsumer(
        'topic',
        loop=asynio.get_running_loop(),
        client_id="experiment_consumer",
        bootstrap_servers=f"server:port",
        enable_auto_commit=False,
    )
    await consumer.start()
    async for msg in consumer:
        print("consumed: ", msg.topic, msg.partition, msg.offset,
                msg.key, msg.value, msg.timestamp)

@app.on_event('startup')
async def app_startup():
    asyncio.create_task(consume())
```
I would expect messages to be printed when published to a topic but nothing gets printed out. As soon as I restart the FastAPI process then the points are being printed. 
I think this may have to do with the async generator used by AIOKafka. Any ideas on what might be going on here?

**Comment by Dustyposa** [comment]
you can try:

```python
@app.on_event('startup')
async def app_startup():
    await consume()
```

**Comment by m4nuC** [comment]
@Dustyposa thanks for the suggestion, I did try this, it results in the application startup hanging as the  `consume` call never terminates(it's an async generator)

**Comment by Dustyposa** [comment]
got it.
You need run it in another process.
Why you want to run it with app in same process？

**Comment by m4nuC** [comment]
it should be possible to work on a single thread, it's what asyncio is for after all. So I'd like to keep multithreading as a last resort

**Comment by insomnes** [comment]
@m4nuC hello! Are you sure that all is fine with your AIOKafkaConsumer setup?
With this async gen emulation I've achieved constant printing:
```
async def async_gen():
    i = 0
    while True:
        await asyncio.sleep(0.5)
        yield i
        i += 1

async def consume():
    async for msg in async_gen():
        print(f"i: {msg}")


app = FastAPI()


@app.on_event('startup')
async def app_startup():
    asyncio.create_task(consume())
```

**Comment by m4nuC** [comment]
@insomnes Thanks for the snippet. I have tried that code have exactly the same symptoms, that is, no printing then I get everything all at once when I restart the FastAPI process. What is even stranger is that If I use the standard python logging library then the printing is indeed constant. Wonder if this could have something to do with gunicorn and possibility multiple fastAPI processes running. Looking into that now.


In dev environment (have not yet tried prod), this is how I start the server:
`exec uvicorn --reload --reload-dir /app --host $HOST --port $PORT --log-level $LOG_LEVEL "$APP_MODULE"`

Could anyone spot something odd ?

**Comment by insomnes** [comment]
@m4nuC hello!
As you've closed your issue I think that you found solution.
Can you explain what was the cause of your problem, please?

**Comment by m4nuC** [comment]
@insomnes It turned out to be the `print` command that was buffering output once inside an async generator. So flushing with `sys.out.flush` on each iteration solve my issue. That said I am not sure why this behavior happens only inside the generator

**Comment by tiangolo** [comment]
Thanks for the help here @Dustyposa and @insomnes  ! 👏  🙇 

Thanks for reporting back and closing the issue @m4nuC  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
